.github/workflows/ci-pipeline.yml

```diff
@@ -17,4 +23,47 @@ jobs:
     - name: "Unit Tests"
       run: docker-compose -f docker-compose-test-unit.yml up --exit-code-from todo-app
     - name: "End to end Tests"
-      run: docker-compose -f docker-compose-test-e2e.yml up --exit-code-from todo-app
\ No newline at end of file
+      run: docker-compose -f docker-compose-test-e2e.yml up --exit-code-from todo-app
+
+  release:
+    name: Release to Production
+    runs-on: ubuntu-latest
+    env:
+      HEROKU_API_KEY: ${{ secrets.HEROKU_API_KEY }}
+    if: ${{ github.ref_name == 'main' }}
+    needs: test
+    steps: 
+    - name: "Checkout code"
+      uses: actions/checkout@v2
+
+    - name: "Create env file"
+      run: |
+        touch .env
+        echo "${{ secrets.SERVER_ENV }}" > .env
+
+    - name: Login to DockerHub
+      uses: docker/login-action@v2
+      with:
+        username: ${{ secrets.DOCKERHUB_USERNAME }}
+        password: ${{ secrets.DOCKERHUB_TOKEN }}
+
+    - name: Docker build
+      run: docker-compose -f docker-compose-prod.yml build
+
+    - name: Tag DockerHub build
+      run: docker tag todo-app:prod thomasjohnstonsoftwire/todo-app:${{ github.sha }}
+
+    - name: DockerHub push
+      run: docker push thomasjohnstonsoftwire/todo-app:${{ github.sha }}
+    
+    - name: Login to Heroku docker
+      run: docker login --username=_ --password=${{ secrets.HEROKU_API_KEY }} registry.heroku.com
```

Comment:
Bit odd to do your `docker login` two different ways (once with an action, ones with a run)

The docker/login-action lets you specify a registry.